Source/OxyPlot/Series/StairStepSeries.cs

```diff
@@ -216,41 +222,146 @@ public override void Render(IRenderContext rc)
                     }
                 };
 
-            // Transform all points to screen coordinates
-            // Render the line when invalid points occur
+            var points = ActualPoints;
+
+            int offset = 0;
+            double xClipMax = double.MaxValue;
+
+            if (IsXMonotonic)
+            {
+                double xClipMin = XAxis.ClipMinimum;
+                xClipMax = XAxis.ClipMaximum;
+
+                WindowStartIndex = UpdateWindowStartIndex(points, point => point.X, xClipMin, WindowStartIndex);
+                offset = WindowStartIndex;
+            }
+
             var linePoints = new List<ScreenPoint>();
             var markerPoints = new List<ScreenPoint>();
-            double previousY = double.NaN;
-            foreach (var point in this.ActualPoints)
+
+            for (int i = offset; i < points.Count;)
             {
-                if (!this.IsValidPoint(point))
+                bool hasValid = FindNextValidSegment(points, i, xClipMax, out int validOffset, out int endOffset);
+                if (!hasValid)
+                    break;
+
+                ScreenPoint transformedPoint = default;
+                DataPoint previousPoint = DataPoint.Undefined;
+                bool xIncreased = false;
+                bool xDecreased = false;
+
+                for (i = validOffset; i < endOffset; ++i)
                 {
-                    renderPoints(linePoints, markerPoints);
-                    linePoints.Clear();
-                    markerPoints.Clear();
-                    previousY = double.NaN;
-                    continue;
+                    var point = points[i];
+
+                    // For performance we want to draw as few lines as possible.  Assuming sane
+                    // axis transformations, as long as point Xs are all either increasing or
+                    // decreasing and point Ys are equal, we can coalesce the horizontal lines and
+                    // omit the vertical lines.
+
+                    xIncreased |= point.X > previousPoint.X;
+                    xDecreased |= point.X < previousPoint.X;
+
+                    if (xIncreased && xDecreased)
+                    {
+                        // Vertical line.
+                        linePoints.Add(Transform(previousPoint));
+
+                        // Horizontal line.
+                        linePoints.Add(Transform(previousPoint));
+
+                        xIncreased = false;
+                        xDecreased = false;
+                    }
```

Comment:
What's the idea here? Should this be projecting two different points rather than the previous point twice?

Does it need to clear both increased and decreased?

Reply:
It's the same point twice because the number of points in the line has to be even to make the horizontal/vertical line styling work.

You're right -- there's a bug in here related to clearing both flags.  I'll fix it.

Reply:
Fixed.